protostar/cheatable_starknet/controllers/expect_call_controller.py

```diff
@@ -0,0 +1,96 @@
+from dataclasses import dataclass
+from typing import TYPE_CHECKING
+
+from protostar.testing.test_environment_exceptions import ExpectedCallException
+from protostar.starknet import CairoOrPythonData, Address
+from protostar.starknet.selector import Selector
+
+if TYPE_CHECKING:
+    from protostar.testing import Hook
+    from protostar.cheatable_starknet.cheatables.cheatable_cached_state import (
+        CheatableCachedState,
+    )
+
+
+@dataclass
+class ExpectedCall:
+    address: Address
+    fn_selector: Selector
+    calldata: CairoOrPythonData
+
+
+def remove_expected_call_external(
+    expected_call: ExpectedCall, cheatable_state: "CheatableCachedState"
+):
+    data_for_address = cheatable_state.expected_contract_calls.get(
+        expected_call.address
+    )
+    if data_for_address is not None:
+        for index, (selector, calldata) in enumerate(data_for_address):
+            if (
+                selector == int(expected_call.fn_selector)
+                and calldata == expected_call.calldata
+            ):
+                del data_for_address[index]
+        if not data_for_address:
+            del cheatable_state.expected_contract_calls[expected_call.address]
+
+
+class ExpectCallController:
+    def __init__(
+        self,
+        test_finish_hook: "Hook",
+        cheatable_state: "CheatableCachedState",
+    ) -> None:
+        self._test_finish_hook = test_finish_hook
+        self._cheatable_state = cheatable_state
+        self._test_finish_hook.on(self.assert_no_expected_calls_left)
```

Comment:
This one is weird, since i think there are no side effects to instantiation of controllers anywhere else. I think this should be exposed as a method and called elsewhere. It's going to work as-is but from my perspective the disctintion between the controller and test control flow is now getting a bit blurry.

Reply:
Discussed offline, as it is the same as in https://github.com/software-mansion/protostar/blob/735ac6cac0a9053348ab33dc8758c2520cb2dc09/protostar/cheatable_starknet/controllers/expect_events_controller.py#L69, we decided to do this separately: https://github.com/software-mansion/protostar/issues/1500